app/services/PageRenderer.scala

```diff
@@ -18,28 +18,29 @@ package services
 
 import javax.inject.{Inject, Singleton}
 import scala.annotation.tailrec
-import models.ocelot.stanzas.{EndStanza, Stanza, Evaluate, DataInput}
+import models.ocelot.stanzas.{EndStanza, VisualStanza, Stanza, Evaluate, DataInput}
 import models.ocelot.{Page, Labels}
 
 @Singleton
 class PageRenderer @Inject() () {
 
   @tailrec
-   private def evaluateStanzas(stanzaId: String, labels: Labels, visualStanzas: Seq[Stanza], seen: Seq[String])
-                              (implicit stanzaMap: Map[String, Stanza]): (Seq[Stanza], Labels, Seq[String], String, Option[DataInput]) =
+   private def evaluateStanzas(stanzaId: String, labels: Labels, visualStanzas: Seq[VisualStanza], seen: Seq[String])
+                              (implicit stanzaMap: Map[String, Stanza]): (Seq[VisualStanza], Labels, Seq[String], String, Option[DataInput]) =
     stanzaMap.get(stanzaId) match {
       case None => (visualStanzas, labels, seen, stanzaId, None)
       case Some(s) => s match {
-        case EndStanza => (visualStanzas :+ EndStanza, labels, seen :+ stanzaId, stanzaId, None)
-        case s: Stanza with DataInput => (visualStanzas :+ s, labels, seen :+ stanzaId, stanzaId, Some(s))
+        case EndStanza => (visualStanzas, labels, seen :+ stanzaId, stanzaId, None)
+        case s: VisualStanza with DataInput => (visualStanzas :+ s, labels, seen :+ stanzaId, stanzaId, Some(s))
         case s: Stanza with Evaluate =>
           val (next, updatedLabels) = s.eval(labels)
           evaluateStanzas(next, updatedLabels, visualStanzas, seen :+ stanzaId)
-        case s: Stanza => evaluateStanzas(s.next.head, labels, visualStanzas :+ s, seen :+ stanzaId)
+        case s: VisualStanza => evaluateStanzas(s.next.head, labels, visualStanzas :+ s, seen :+ stanzaId)
+        case s: Stanza => evaluateStanzas(s.next.head, labels, visualStanzas, seen :+ stanzaId) // Never should happen catch all
```

Comment:
Should we log an error here

Reply:
not an error, think the comment I added is misleading. It allows for stanzas which are not Visual, DataInput or Evaluate to exist in the page but be ignored. They are included in the "seen" list only. Until recently the only candidate was the CalculationStanza.

Reply:
And now Calculation implements eval.